### PR TITLE
drivers/net/ppp: fix net_buf corruption when truncating packet

### PR DESCRIPTION
**Pull request by jimparis**
To truncate the FCS bytes off the end of the PPP data, the last fragment's length is reduced by 2.  However, the last frag may contain only 0 or 1 bytes of data, in which case data needs to be removed from the parent instead.  If the parent is not full, the extra fragment also needs to be removed.

Without this patch, receiving a PPP packet that is a multiple of CONFIG_NET_BUF_DATA_SIZE will result in the last fragment's buffer length wrapping around to 65534.

### CODE REVIEW

**Review by jukkar — APPROVED** [review]

File: drivers/net/ppp.c
Comment (mniestroj) [inline drivers/net/ppp.c:442]: This seems too complicated to me. First of all, make sure that this implementation details are in a separate "helper" function. Second, what about making it a bit more generic:

```
/*
 * Truncate 'len' bytes from tail of net_pkt, assuming that requested remove
 * area spans maximum of 2 fragments.
 */
static void net_pkt_truncate_tail(struct net_pkt *pkt, size_t len)
{
	struct net_buf *buf = pkt->frags;
	struct net_buf *prev = NULL;

	/* find two last fragments */
	while (buf->frags) {
		prev = buf;
		buf = buf->frags;
	}

	/* truncate last packet */
	if (buf->len >= len) {
		buf->len -= len;
		return;
	}

	/* remove last packet */
	net_buf_frag_del(prev, buf);

	/* truncate "new" last packet */
	prev->len -= len;
}
```

Comment (jimparis) [reply drivers/net/ppp.c:442]: Agreed on making a helper.
That suggested code appears to have at least two bugs though:
- If `buf->len < len`, you'll remove the full `len` from `prev`, where you should have removed `len - buf->len`
- If `prev` has less than `len` bytes, you'll underflow
I'm not sure that the helper should be limited to a max of 2 fragments, either.  We can't walk the fragment list backwards, but you could just loop back to the top and scan the list again -- it's slow but would only happen in the case that would otherwise not work at all.

Comment (mniestroj) [reply drivers/net/ppp.c:442]: >If buf->len < len, you'll remove the full len from prev, where you should have removed len - buf->len

Yeah, I didn't test it. You are right.

>If prev has less than len bytes, you'll underflow

I don't think this is possible in this file context. I was assuming that this will be a static function used only in PPP context.

>I'm not sure that the helper should be limited to a max of 2 fragments, either. We can't walk the fragment list backwards, but you could just loop back to the top and scan the list again -- it's slow but would only happen in the case that would otherwise not work at all.

I was thinking about that. The function that I proposed is a compromise between begin generic, simple and most importatnly solve the task of removing small portion (less than a net_buf can contain) od data from net_buf fragments.

We can implement a generic function that will work with any set of parameters and any type of fragment list (also such that contains only one byte in a single net_buf fragment). My idea was to reverse net_buf fragment list by manipulating `net_buf->frags`, then calling something like `net_pkt_pull` API and then reverse net_buf fragment list once again. This would allow to keep `O(n)`.

Comment (jimparis) [reply drivers/net/ppp.c:442]: I think we can do it more easily in `O(n)`: walk fragments and count off `net_pkt_get_len(pkt) - len` bytes, then truncate wherever we run out, and delete fragments after that.

Comment (pfalcon) [reply drivers/net/ppp.c:442]: @jimparis, @mniestroj: What's the status of this? @mniestroj, if you think that factoring out that code as a separate static function is really required, please give -1 until it's done. Otherwise, I'm keen to give my +1.

But also a friendly reminder that without tests for such cases, we'll be drowned in issues like this (newer changes regress/mis-interact with previous changes, etc.). (Btw, that would be a good "pro" for having a generic/non-static function like that, then we can unit-test it separately, which should be easier than "integrated" test with ppp-specific handling.)

Comment (mniestroj) [reply drivers/net/ppp.c:442]: I think that this implementation is too complicated, but I won't block this PR (by -1) because of it. It is tricky to follow so many if-statements (there are simply too many branches) and there is no test for it.

I like @jimparis idea of:

>I think we can do it more easily in O(n): walk fragments and count off net_pkt_get_len(pkt) - len bytes, then truncate wherever we run out, and delete fragments after that.

together with @pfalcon suggestion of:

>Btw, that would be a good "pro" for having a generic/non-static function like that, then we can unit-test it separately, which should be easier than "integrated" test with ppp-specific handling.

which should give something like:

```
int net_pkt_remove_tail(struct net_pkt *pkt, size_t length)
{
	struct net_buf *buf = pkt->buffer;
	size_t pkt_len = net_pkt_get_len(pkt);

	if (pkt_len < length) {
		return -EINVAL;
	}

	pkt_len -= length;

	while (buf) {
		if (buf->len >= pkt_len) {
			buf->len = pkt_len;

			if (buf->frags) {
				net_pkt_frag_unref(buf->frags);
				buf->frags = NULL;
			}

			break;
		}

		buf = buf->frags;
		pkt_len -= buf->len;
	}

	return 0;
}
```

and easy way to add unit tests for it.

Comment (pfalcon) [reply drivers/net/ppp.c:442]: > I think that this implementation is too complicated, but I won't block this PR (by -1) because of it. 

Right, it's a fix for obvious bug after all. Ping @jimparis, can you please do suggested refactorings, to make sure this fits into 2.6 (the freeze is soon AFAIK).

Comment (jukkar) [reply drivers/net/ppp.c:442]: > 		if (buf->len >= pkt_len) {
> 			buf->len = pkt_len;

Hmm, this does not look right. The `buf->len` is just the length of this small fragment, and we are comparing it to the total length of the pkt. I think we should compare to cumulative length of the pkt so far.
Also, setting `buf->len = pkt_len;` would set the length of this fragment to the total length of the pkt which would be wrong if there are multiple fragments in this chain.

Comment (mniestroj) [reply drivers/net/ppp.c:442]: @jukkar Perhaps `pkt_len` is a bad variable name then. Maybe `remaining_len` would better match implemented logic. Then:
```
if (buf->len >= remaining_len) {
  	buf->len = remaining_len;
```
Notice that after entering `if (buf->len >= remaining_len) {` we already are at the last buffer that should remain, all the other buffers should be dropped. So `buf->len = remaining_len;` is either reducing length of last buffer or leaving it as is.

>Also, setting buf->len = pkt_len; would set the length of this fragment to the total length of the pkt which would be wrong if there are multiple fragments in this chain.

That is why `remaining_len` would be probably better name, as we decrease that variable in a loop by length of buffers in packet. So that variable actually represents how many bytes should still be left in packet.

Comment (jukkar) [reply drivers/net/ppp.c:442]: > Maybe `remaining_len` would better match implemented logic

Ack to that :)

**Comment by cfriedt** [comment]
@rlubos - do you want to take a look at this as well? It sounds as though this is a bug, but there is not much in the way of test coverages.

Personally, I think that HDLC would be good to be decoupled from PPP, as it would become more unit-testable in that case and it could also be used in kind of weird places to provide a somewhat reliable link layer.

I did something like that a while ago - created the inverse of a USB to Serial adapter (a serial to USB adapter). That actually went into a product already.

https://github.com/zephyrproject-rtos/zephyr/compare/main...cfriedt:hdlc?expand=1

**Comment by rlubos** [comment]
> @rlubos - do you want to take a look at this as well? It sounds as though this is a bug, but there is not much in the way of test coverages.
> 
> Personally, I think that HDLC would be good to be decoupled from PPP, as it would become more unit-testable in that case and it could also be used in kind of weird places to provide a somewhat reliable link layer.
> 
> I did something like that a while ago - created the inverse of a USB to Serial adapter (a serial to USB adapter). That actually went into a product already.
> 
> https://github.com/zephyrproject-rtos/zephyr/compare/main...cfriedt:hdlc?expand=1

Well, if HDLC could be decoupled from PPP, then I think it's a good idea, we just need a volunteer. 

As for the PR itself, I think we should rather focus on https://github.com/zephyrproject-rtos/zephyr/pull/36972, as it seems to address the issue in a neater way (i. e. it covers the concerns raised here).

**Comment by cfriedt** [comment]
@jimparis - if you're ok with the solution that @rlubos has proposed in #36972 then please close this PR. In any case, thank you for all of the work you did to bring this to the maintainers' attention.

**Comment by cfriedt** [comment]
Fix was duplicated in  #36972 but  that one included unit tests.

**Pull request closed without merging**
